abi-versions/vNEXT/README.md

```diff
@@ -0,0 +1,919 @@
+# Proxy-Wasm vNEXT ABI specification
+
+# Functions implemented in the Wasm module
+
+All functions implemented in the Wasm module, other than the integration and memory management
+functions, include context identifier (`context_id`) as the first parameter, which should be used to
+distinguish between different contexts.
+
+
+## Integration
+
+### `_start`
+
+* params:
+  - none
+* returns:
+  - none
+
+Start function which is called when the module is loaded and initialized. This can be used by SDKs
+to setup and/or initialize state, but no proxy_ functions can be used at that point yet.
+
+
+### `proxy_abi_version_X_Y_Z`
+
+* params:
+  - none
+* returns:
+  - none
+
+Exports ABI version (vX.Y.Z) this module was compiled against.
+
+Considered alternatives: Ideally, this would be exported as a global variable, but existing
+toolchains have trouble with that (some don’t support global variables at all, some incorrectly
+export language-specific and not Wasm-generic variables).
+
+
+## Memory management
+
+### `proxy_on_memory_allocate`
+
+* params:
+  - `i32 (size_t) memory_size`
+* returns:
+  - `i32 (void*) allocated_ptr`
+
+Allocates memory using in-VM memory allocator and returns it to the host.
+
+
+## Module lifecycle
+
+### `proxy_on_context_create`
+
+* params:
+  - `i32 (uint32_t) context_id`
+  - `i32 (uint32_t) parent_context_id`
+* returns:
+  - none
```

Comment:
I had  a [discussion in the architecture doc](https://docs.google.com/document/d/1YPbxuEIPwKsW4Ft73diHRBTUNbYIlrQHNnS9dXdn778/edit?disco=AAAAEHXWueQ) with @jplevyak on this function, that I would  like to get the opinion of the reviewers on.

i think it would be useful to have a way to allow runtimes use a void* coming from the runtime as their context id, for the following reasons:
- this id is used to retrieve the context on every request. this means that the runtime essentially has to store these in a map<id, context> for fast access. some runtimes (tiny-go comes to  mind) may not have an efficient map implementation.
- it will make runtimes simpler\easier to write

Reply:
You mean returning ID that's allocated by the Wasm VM, that you could (ab)use to return memory address of the context struct? Sure, that's possible, but how many languages would actually benefit from it? Also, it seems that you'd need to keep track of allocated IDs in Wasm VM, to make sure that the `void*` ID in the callback still points to a valid context, so I'm not sure how much performance gain you'd get from this in practice.

Reply:
i can see the implementations of most runtimes simplefied by removing the need of a map. essentially every language with a notion of a pointer (c++,rust, assembly-script,tinygo..)

Why would we need keep track of the `void*` ID inside the VM? could it be used with a `proxy_*` callback after `proxy_on_delete`  was called?

Reply:
From a pure perf perspective I agree with @yuval-k that this would likely work and be usable in many languages. I don't feel strongly if we try to do this now or leave this as a possible post MVP optimization. Perhaps we should start opening issues against this repo?

Reply:
Sure, it's definitely better from the perf point of view, since you're avoiding map lookup for each callback. Having said that, I'm feeling a bit iffy about having pointer to the context maintained outside of the Wasm VM.

How about we allow Wasm VM to return module-assigned ID that would be used in subsequent calls? This way, you could (ab)use this feature and assign the ID corresponding to the memory address. Would that work for you?

Reply:
this sounds good to me